### PR TITLE
Implemenation LDAP client for search query

### PR DESCRIPTION
**Pull request by ShatalovYaroslav**
Purpose of this task is to allow to query a LDAP server from a task in order to retrieve some information about an user, like his email address, groups, Projects, etc.

The LDAP query task will have next view:

// Please fill variables
ldapUrl = "ldap://localhost:389"
ldapSearchBase = "dc=yourOrganization,dc=com"
ldapSearchFilter = "(objectclass=*)"
ldapSelectedAttributes = "attributeName1,attributeName2"

// Query a LDAP server requires authentication for connection
variables.put("ldapUrl", ldapUrl);
variables.put("ldapSearchBase", ldapSearchBase);
variables.put("ldapSearchFilter", ldapSearchFilter);
variables.put("ldapSelectedAttributes", ldapSelectedAttributes);

import org.ow2.proactive.addons.ldap_query.LDAPClient;
LDAPClient ldapClient  = new LDAPClient(variables, credentials);
String ldapResult = ldapClient.searchQueryLDAP();
result = ldapResult;

### CODE REVIEW

**Review by laurianed — APPROVED** [review]

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.addons.ldap_query;/*
+                                            * ProActive Parallel Suite(TM):
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java]: Header appears twice.

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        this.ldapUrl = ldapUrl;
+        this.ldapUsername = ldapUsername;
+        this.ldapPassword = ldapPassword;
+        this.ldapSearchBase = ldapSearchBase;
+        this.ldapSearchFilter = ldapSearchFilter;
+        this.ldapSelectedAttributes = ldapSelectedAttributes;
+    }
+
+    public LDAPClient(Map<String, Serializable> actualTaskVariables, Map<String, Serializable> credentials) {
+        List<String> taskVariablesList = new LinkedList<>();
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java]: Initialize with capacity? or ImmutableList.of(...)?

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        taskVariablesList.add(ARG_URL);
+        taskVariablesList.add(ARG_USERNAME);
+        taskVariablesList.add(ARG_PASSWORD);
+        taskVariablesList.add(ARG_SEARCH_BASE);
+        taskVariablesList.add(ARG_SEARCH_FILTER);
+        taskVariablesList.add(ARG_SELECTED_ATTRIBUTES);
+
+        Map mapWithVariables;
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java]: Define generic types: Map<..,..>

File: src/main/java/org/ow2/proactive/addons/ldap_query/model/ErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * @author ActiveEon Team
+ * @since 1/20/2017
+ */
+
+@Getter
+@NoArgsConstructor
+public class ErrorResponse extends Response {
+    private String errorMessage;
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/model/ErrorResponse.java:40]: final?

File: src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Map;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+
+@Getter
+@NoArgsConstructor
+public class LDAPResponse extends Response {
+    private List<Map<String, String>> attributes;
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java:38]: final?

File: src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    public LDAPResponse(String status, List<Map<String, String>> attributes) {
+        this.status = status;
+        this.attributes = attributes;
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java:42]: Create an immutable copy? ImmutableList.copyOf?

File: src/main/java/org/ow2/proactive/addons/ldap_query/model/Response.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * @author ActiveEon Team
+ * @since 1/20/2017
+ */
+
+@Getter
+@NoArgsConstructor
+@ToString
+public abstract class Response {
+    protected String status;
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/model/Response.java:44]: final?

File: src/test/java/org/ow2/proactive/addons/ldap_query/LDAPClientTest.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.addons.ldap_query;/*
+                                            * ProActive Parallel Suite(TM):
```
Comment (laurentpellegrino) [inline src/test/java/org/ow2/proactive/addons/ldap_query/LDAPClientTest.java]: Duplicate header?

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        return splittedAttr;
+    }
+
+    public String searchQueryLDAP() {
+        NamingEnumeration results = null;
+        ObjectMapper mapper = new ObjectMapper();
+        Response response;
+        boolean allAttrs = false;
+        String resultOutput = new String();
+        List attributesList = new LinkedList();
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java]: Define generic type: List<...>

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+            ldapConnection = LDAPConnectionUtility.connect(ldapUrl, ldapUsername, ldapPassword);
+            SearchControls controls = new SearchControls();
+            controls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+            results = ldapConnection.search(ldapSearchBase, ldapSearchFilter, controls);
+
+            //iterate throw all attributes in the result of search query
+            while (results.hasMore()) {
+                SearchResult searchResult = (SearchResult) results.next();
+                Attributes attributes = searchResult.getAttributes();
+                Map<String, String> attrMap = new HashMap();
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java]: Use diamonds: new HashMap<>();

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPConnectionUtility.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+ * @since 1/25/2017
+ */
+public class LDAPConnectionUtility {
+
+    private final static String LDAP_FACTORY = "com.sun.jndi.ldap.LdapCtxFactory";
+
+    private final static String SECURITY_AUTHENTICATION_METHOD = "simple";
+
+    public static DirContext connect(String ldapUrl, String ldapUsername, String ldapPassword) throws NamingException {
+        Hashtable env = new Hashtable();
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPConnectionUtility.java]: Generic type def.

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                }
+            }
+            response = new LDAPResponse("Ok", attributesList);
+        } catch (Exception e) {
+            response = new ErrorResponse("Error", e.toString());
+        } finally {
+            if (results != null) {
+                try {
+                    results.close();
+                } catch (Exception e) {
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java:180]: Please do something with the exception. As least, log or print to standard err if the purpose is to be used from a ProActive Task.

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+                }
+            }
+            if (ldapConnection != null) {
+                try {
+                    ldapConnection.close();
+                } catch (Exception e) {
+                }
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java:189]: Same as before.

File: src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+        }
+    }
+
+    private void setLdapClientFields(Map<String, Serializable> taskVariablesMap, String variableName) {
+        if (!taskVariablesMap.containsKey(variableName)) {
+            throw new IllegalArgumentException("The missed argument for LDAPClient, variable name: " + variableName);
+        }
+        String varValue = getAsString(taskVariablesMap, variableName);
+        try {
+            this.getClass().getDeclaredField(variableName).set(this, varValue);
```
Comment (laurentpellegrino) [inline src/main/java/org/ow2/proactive/addons/ldap_query/LDAPClient.java]: I don't like introspection for such a use case. It's easy to break in the future and it has impacts on performance.

You have checked just before that the expected variable exists. Why do you need to copy fields to dedicated fields. Is it not possible to simply query the map when required?

**Review by laurentpellegrino — CHANGES_REQUESTED** [review]

File: src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java (hunk 5 above)
Comment (ShatalovYaroslav) [reply src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java:38]: In fact I wanted to make it final. But when it is final, it should be deleted default noArgsConstructor.
Then when I parse to json object with jackson, receive the next error:
 "can not instantiate from JSON object (missing default constructor or creator, or perhaps need to add/enable type information?)"
So jackson needs default constructor.

Comment (laurentpellegrino) [reply src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java:38]: Ah ok, sorry

File: src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java (hunk 6 above)
Comment (ShatalovYaroslav) [reply src/main/java/org/ow2/proactive/addons/ldap_query/model/LDAPResponse.java:42]: The same issue for parsing from jackson. As guava ImmutableList doesn't have default constructor.
I've tried to add jackson-datatype-guava dependency. But still have error while parsing.

**Review by laurentpellegrino — APPROVED** [review]

**Pull request merged**